### PR TITLE
ADM-760: [Frontend] fix pipeline steps&branches lost in metrics page

### PR DESCRIPTION
**Pull request by gabralia**
## Summary

https://dorametrics.atlassian.net/browse/ADM-760

1. refactor real done code
2. fix pipeline steps&branches lost in metrics page

### CODE REVIEW

File: frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx [hunk 1]
```diff
@@ -11,12 +11,11 @@ interface Props {
   value: string;
   id: number;
   onGetSteps?: (pipelineName: string) => void;
-  step?: string;
   onUpDatePipeline: (id: number, label: string, value: string) => void;
 }
 
 /* istanbul ignore next */
```
Comment (SimonTal) [inline frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx:17]: Try to remove istanbul ignore

**Review by SimonTal — APPROVED** [review]
Approved

File: frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx (hunk 1 above)
Comment (SimonTal) [reply frontend/src/containers/MetricsStep/DeploymentFrequencySettings/SingleSelection/index.tsx:17]: Currently skip due to bug bash

File: frontend/__tests__/utils/Util.test.tsx [hunk 2]
```diff
@@ -129,3 +131,61 @@ describe('filterAndMapCycleTimeSettings function', () => {
...
+      { column: 'TODO', status: 'ToDo', value: 'TODO' },
+      { column: 'TODO', status: 'Backlog', value: 'TODO' },
+      { column: 'IN DEV', status: 'InDev', value: 'IN DEV' },
+      { column: 'IN DEV', status: 'Doing', value: 'Done' },
+      { column: 'DONE', status: 'DONE', value: 'Done' },
+    ];
+
+    const result = getRealDoneStatus(MOCK_CYCLE_TIME_SETTING, CYCLE_TIME_SETTINGS_TYPES.BY_COLUMN, ['Doing']);
+
+    expect(result).toEqual(['Doing']);
```
Comment (guzhongren) [inline frontend/__tests__/utils/Util.test.tsx:175]: i think the expectation does not match the test description

File: frontend/__tests__/context/configSlice.test.ts [hunk 3]
```diff
@@ -100,4 +102,9 @@ describe('config reducer', () => {
 
     expect(config.warningMessage).toEqual(CONFIG_PAGE_VERIFY_IMPORT_ERROR_MESSAGE);
   });
+
+  it('should return empty when steps not exists in pipelineList', () => {
```
Comment (guzhongren) [inline frontend/__tests__/context/configSlice.test.ts:106]: when is an action, could u replace it with `given`

File: frontend/__tests__/hooks/useMetricsStepValidationCheckContext.test.tsx [hunk 4]
```diff
@@ -35,78 +17,28 @@ describe('useMetricsStepValidationCheckContext', () => {
...
-    expect(result.current?.isPipelineValid(DEPLOYMENT_FREQUENCY_SETTINGS)).toBe(false);
-    expect(result.current?.isPipelineValid(LEAD_TIME_FOR_CHANGES)).toBe(false);
     expect(
-      result.current?.getDuplicatedPipeLineIds([{ id: 1, organization: '', pipelineName: '', step: '', branches: [] }]),
+      result.current?.getDuplicatedPipeLineIds([{ id: 0, organization: '', pipelineName: '', step: '', branches: [] }]),
     ).toEqual([]);
   });
 
-  it('should return useMetricsStepValidationCheckContext correctly ', () => {
+  it('should return empty id when params is empty array', () => {
```
Comment (guzhongren) [inline frontend/__tests__/hooks/useMetricsStepValidationCheckContext.test.tsx:28]: when -> given

File: frontend/__tests__/utils/Util.test.tsx [hunk 5]
```diff
@@ -129,3 +131,61 @@ describe('filterAndMapCycleTimeSettings function', () => {
     expect(result).toEqual(expected);
   });
 });
+
+describe('getRealDoneStatus', () => {
+  it('should return selected done status when cycle time settings only have one done value and type is by column', () => {
```
Comment (guzhongren) [inline frontend/__tests__/utils/Util.test.tsx:136]: when -> given in the all files

File: frontend/__tests__/utils/Util.test.tsx [hunk 6]
```diff
@@ -129,3 +131,61 @@ describe('filterAndMapCycleTimeSettings function', () => {
...
+  });
+
+  it('should return selected done status when cycle time settings only have one done value and type is by status', () => {
+    const MOCK_CYCLE_TIME_SETTING = [
+      { column: 'TODO', status: 'ToDo', value: 'TODO' },
+      { column: 'TODO', status: 'Backlog', value: 'TODO' },
+      { column: 'IN DEV', status: 'InDev', value: 'IN DEV' },
+      { column: 'IN DEV', status: 'Doing', value: 'IN DEV' },
+      { column: 'DONE', status: 'DONE', value: 'Done' },
+    ];
```
Comment (guzhongren) [inline frontend/__tests__/utils/Util.test.tsx:157]: duplicated code

File: frontend/__tests__/utils/Util.test.tsx [hunk 7]
```diff
@@ -129,3 +131,61 @@ describe('filterAndMapCycleTimeSettings function', () => {
...
+
+    const result = getRealDoneStatus(MOCK_CYCLE_TIME_SETTING, CYCLE_TIME_SETTINGS_TYPES.BY_STATUS, []);
+
+    expect(result).toEqual(['DONE']);
```
Comment (guzhongren) [inline frontend/__tests__/utils/Util.test.tsx:161]: relace 'DONE' with the constant or fixture

File: frontend/__tests__/utils/Util.test.tsx [hunk 8]
```diff
@@ -129,3 +131,61 @@ describe('filterAndMapCycleTimeSettings function', () => {
...
+  });
+
+  it('should return selected done status when cycle time settings type is by column', () => {
+    const MOCK_CYCLE_TIME_SETTING = [
+      { column: 'TODO', status: 'ToDo', value: 'TODO' },
+      { column: 'TODO', status: 'Backlog', value: 'TODO' },
+      { column: 'IN DEV', status: 'InDev', value: 'IN DEV' },
+      { column: 'IN DEV', status: 'Doing', value: 'Done' },
+      { column: 'DONE', status: 'DONE', value: 'Done' },
+    ];
```
Comment (guzhongren) [inline frontend/__tests__/utils/Util.test.tsx:185]: duplicated code

**Review by guzhongren — COMMENTED** [review]
Change with the comments

**Comment by SimonTal** [comment]
Deploy with comments due to bug bash

**Pull request merged**